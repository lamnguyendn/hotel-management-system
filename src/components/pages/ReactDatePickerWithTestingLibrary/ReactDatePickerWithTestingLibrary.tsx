import React, { useState } from 'react';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';

const maxDate = new Date();
maxDate.setDate(maxDate.getDate() + 1);

const Input: React.FC<any> = (props) => <input {...props} />;

export const ReactDatePickerWithTestingLibrary = (): JSX.Element => {
  const [value, setValue] = useState(new Date());

  function handleChange(value: any) {
    //
    setValue(value);
  }

  return (
    <div className="App">
      <label htmlFor="from">From:</label>
      <DatePicker
        id="date-picker"
        onChange={handleChange}
        maxDate={maxDate}
        selected={value}
        strictParsing
        dateFormat="yyyy/MM/dd"
        // customInput={(props: any) => <Input {...props} data-testid="date-picker" />}
      />
    </div>
  );
};
