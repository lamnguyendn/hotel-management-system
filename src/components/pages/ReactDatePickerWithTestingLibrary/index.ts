export { ReactDatePickerWithTestingLibrary as ReactDatePickerWithTestingLibraryPage } from './ReactDatePickerWithTestingLibrary';
