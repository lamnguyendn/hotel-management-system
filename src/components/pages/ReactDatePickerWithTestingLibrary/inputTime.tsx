/* eslint-disable react/destructuring-assignment */
import PropTypes from 'prop-types';
import React from 'react';

const propTypes = {
  onChange: PropTypes.func,
  date: PropTypes.instanceOf(Date),
  timeString: PropTypes.string,
  timeInputLabel: PropTypes.string,
  customTimeInput: PropTypes.element,
};

export default class inputTime extends React.Component<any, any> {
  constructor(props: any) {
    super(props);

    this.state = {
      time: this.props.timeString,
    };
  }

  static getDerivedStateFromProps(props: { timeString: any }, state: { time: never }) {
    if (props.timeString !== state.time) {
      return {
        time: props.timeString,
      };
    }

    // Return null to indicate no change to state.
    return null;
  }

  onTimeChange = (time: { split: (arg0: string) => number[] }) => {
    this.setState({ time });
    const date = new Date();
    date.setHours(time.split(':')[0]);
    date.setMinutes(time.split(':')[1]);
    this.props.onChange(date);
  };

  renderTimeInput = () => {
    const { time } = this.state;
    const { date, timeString, customTimeInput } = this.props;

    if (customTimeInput) {
      return React.cloneElement(customTimeInput, {
        date,
        value: time,
        onChange: this.onTimeChange,
      });
    }

    return (
      <input
        type="time"
        className="react-datepicker-time__input"
        placeholder="Time"
        name="time-input"
        required
        value={time}
        onChange={(ev) => {
          this.onTimeChange(ev.target.value || timeString);
        }}
      />
    );
  };

  render() {
    return (
      <div className="react-datepicker__input-time-container">
        <div className="react-datepicker-time__caption">{this.props.timeInputLabel}</div>
        <div className="react-datepicker-time__input-container">
          <div className="react-datepicker-time__input">{this.renderTimeInput()}</div>
        </div>
      </div>
    );
  }
}
