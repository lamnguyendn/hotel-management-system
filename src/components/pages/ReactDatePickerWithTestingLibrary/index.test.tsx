import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import * as React from 'react';
import { ReactDatePickerWithTestingLibrary } from './ReactDatePickerWithTestingLibrary';

afterEach(cleanup);

describe('<MyComponent/>', () => {
  it('should render correctly', () => {
    const component = render(<ReactDatePickerWithTestingLibrary />);
    expect(component).toBeTruthy();
  });

  it('should allow me to enter text into the text box', async () => {
    const { container, getByText } = render(<ReactDatePickerWithTestingLibrary />);
    const label = getByText('From:');
    expect(container).toMatchSnapshot();
    const textbox = container.querySelector('input[id="date-picker"]');
    // fireEvent.click(label);
    fireEvent.change(textbox as HTMLElement, { target: { value: '2022/01/01' } });
    // userEvent.type(textbox as HTMLElement, '2022/01/01');
    // fireEvent.blur(textbox as HTMLElement);
    fireEvent.click(label);

    await waitFor(() => expect((textbox as HTMLElement).getAttribute('value')).toBeEmpty());
  });
});
