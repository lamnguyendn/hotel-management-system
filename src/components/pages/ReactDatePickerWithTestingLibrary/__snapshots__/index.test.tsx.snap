// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MyComponent/> should allow me to enter text into the text box 1`] = `
<div>
  <div
    class="App"
  >
    <label
      for="from"
    >
      From:
    </label>
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container"
      >
        <input
          class=""
          id="date-picker"
          type="text"
          value="2021/01/18"
        />
      </div>
    </div>
  </div>
</div>
`;
